src/NUnitFramework/framework/Constraints/DictionaryContainsKeyConstraint.cs

```diff
@@ -65,32 +71,138 @@ public override string Description
+                foreach (object obj in dictionary.Keys)
+                    if (ItemsEqual(obj, Expected))
+                        return true;
+
+                return false;
+            }
 
-            foreach (object obj in dictionary.Keys)
-                if (ItemsEqual(obj, Expected))
-                    return true;
+            var method = GetContainsKeyMethod(actual);
+            if (method != null)
+                return (bool)method.Invoke(actual, new[] { Expected });
 
-            return false;
+            throw new ArgumentException("Not a collection supporting ContainsKey method.");
         }
 
         /// <summary>
         /// Flag the constraint to use the supplied predicate function
         /// </summary>
         /// <param name="comparison">The comparison function to use.</param>
         /// <returns>Self.</returns>
+        [Obsolete("Deprecated, use Does.ContainKey")]
         public DictionaryContainsKeyConstraint Using<TCollectionType, TMemberType>(Func<TCollectionType, TMemberType, bool> comparison)
         {
             // reverse the order of the arguments to match expectations of PredicateEqualityComparer
             Func<TMemberType, TCollectionType, bool> invertedComparison = (actual, expected) => comparison.Invoke(expected, actual);
 
+            _deprecatedFlag = true;
             base.Using(EqualityAdapter.For(invertedComparison));
             return this;
         }
+
+        /// <summary>
+        /// Flag the constraint to use the supplied Comparison object.
+        /// </summary>
+        /// <param name="comparer">The IComparer object to use.</param>
+        /// <returns>Self.</returns>
+        [Obsolete("Deprecated, use Does.ContainKey")]
+        public new CollectionItemsEqualConstraint Using<T>(Comparison<T> comparer)
+        {
+            _deprecatedFlag = true;
+            return base.Using(comparer);
+        }
+
+        /// <summary>
+        /// Checks if the key is contained in a "keyed item container".
+        /// </summary>
+        /// <param name="actual">Keyed container.</param>
+        /// <returns>method to call.</returns>
+        private MethodInfo GetContainsKeyMethod(object actual)
+        {
+            if (actual == null) throw new ArgumentNullException(nameof(actual));
+            var instanceType = actual.GetType();
+
+            var method = FindContainsKeyMethod(instanceType)
+                         ?? instanceType
+                            .GetInterfaces()
+                            .Concat(GetBaseTypes(instanceType))
+                            .Select(FindContainsKeyMethod)
+                            .FirstOrDefault(m => m != null);
+
+            return method;
+        }
+
+        /// <summary>
+        /// Looks for a base type that implements ContainsKey method
+        /// </summary>
+        /// <param name="type">Type to look for method ContainsKey</param>
+        /// <returns>Returns the method to call.</returns>
```

Comment:
Before I nitpick at this one too, I should point out that leaving the private methods without XML docs is also a good option.

The summary makes it look like it only searches base types, but it also searches the most derived type as well as all implemented interfaces. Then you'd say, "... that implements **a** \<c>ContainsKey\</c> method." (Including the indefinite article, the `<c>`, and the end mark.

The `<param>` has some awkward wording. You can say, "The type to search for X," because it makes sense to say, "Search the type for X." You can't say, "The type to look for X," because it doesn't make sense to say, "Look the type for X." 'Look' is an [intransitive](https://en.wikipedia.org/wiki/Intransitive_verb) verb.
Better would be, "The type which will be searched for a \<c>ContainsKey\</c> method." That way you're also describing the effect the parameter has, not just what the parameter *is*.